lib/semantics/resolve-names-utils.cc

```diff
@@ -286,4 +288,356 @@ Bound ArraySpecAnalyzer::GetBound(const parser::SpecificationExpr &x) {
   return Bound{std::move(expr)};
 }
 
+// If SAVE is set on src, set it on all members of dst
+static void PropagateSaveAttr(
+    const EquivalenceObject &src, EquivalenceSet &dst) {
+  if (src.symbol.attrs().test(Attr::SAVE)) {
+    for (auto &obj : dst) {
+      obj.symbol.attrs().set(Attr::SAVE);
+    }
+  }
+}
+static void PropagateSaveAttr(const EquivalenceSet &src, EquivalenceSet &dst) {
+  if (!src.empty()) {
+    PropagateSaveAttr(src.front(), dst);
+  }
+}
+
+void EquivalenceSets::AddToSet(const parser::Designator &designator) {
+  if (CheckDesignator(designator)) {
+    Symbol &symbol{*currObject_.symbol};
+    if (!currSet_.empty()) {
+      // check this symbol against first of set for compatibility
+      Symbol &first{currSet_.front().symbol};
+      CheckCanEquivalence(designator.source, first, symbol) &&
+          CheckCanEquivalence(designator.source, symbol, first);
+    }
+    auto subscripts{currObject_.subscripts};
+    if (subscripts.empty() && symbol.IsObjectArray()) {
+      // record a whole array as its first element
+      for (const ShapeSpec &spec : symbol.get<ObjectEntityDetails>().shape()) {
+        if (auto &lbound{spec.lbound().GetExplicit()}) {
+          if (auto subscript{evaluate::ToInt64(*lbound)}) {
+            subscripts.push_back(*subscript);
+          }
+        }
+      }
+    }
+    currSet_.emplace_back(symbol, subscripts);
+    PropagateSaveAttr(currSet_.back(), currSet_);
+  }
+  currObject_ = {};
+}
+
+void EquivalenceSets::FinishSet(const parser::CharBlock &source) {
+  std::set<std::size_t> existing;  // indices of sets intersecting this one
+  for (auto &obj : currSet_) {
+    auto it{objectToSet_.find(obj)};
+    if (it != objectToSet_.end()) {
+      existing.insert(it->second);  // symbol already in this set
+    }
+  }
+  if (existing.empty()) {
+    sets_.push_back({});  // create a new equivalence set
+    MergeInto(source, currSet_, sets_.size() - 1);
+  } else {
+    auto it{existing.begin()};
+    std::size_t dstIndex{*it};
+    MergeInto(source, currSet_, dstIndex);
+    while (++it != existing.end()) {
+      MergeInto(source, sets_[*it], dstIndex);
+    }
+  }
+  currSet_.clear();
+}
+
+// Report an error if sym1 and sym2 cannot be in the same equivalence set.
+bool EquivalenceSets::CheckCanEquivalence(
+    const parser::CharBlock &source, const Symbol &sym1, const Symbol &sym2) {
+  parser::MessageFixedText msg{"", 0};
```

Comment:
Does default construction not work for `MessageFixedText`?  If not, it might be better to add a default constructor to that class and use it.

Reply:
No, it doesn't have a default constructor. I think that using optional is a better way to do it anyway, so I'll change it to that.